File: mns-core/src/main/java/cz/fb/manaus/core/category/categorizer/SelectionRegexpCategorizer.java
package cz.fb.manaus.core.category.categorizer;

import cz.fb.manaus.core.category.BetCoverage;
import cz.fb.manaus.core.model.SettledBet;
import org.springframework.stereotype.Component;

import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

import static java.util.stream.Collectors.toSet;

@Component
public class SelectionRegexpCategorizer extends AbstractRegexpResolver implements SettledBetCategorizer {

    public static final Map<String, Pattern> SELECTION_MAP = Map.of(
            "draw", compile("^The\\s+Draw$"),
            "yes", compile("^Yes$"),
            "no", compile("^No$"));

    public SelectionRegexpCategorizer() {
        super("selectionRegexp_");
    }

    @Override
    public Set<String> getCategories(SettledBet settledBet, BetCoverage coverage) {
        String selectionName = settledBet.getSelectionName();
        return getCategories(selectionName);
    }

    Set<String> getCategories(String selectionName) {
        Set<String> selectionBased = getCategories(selectionName, SELECTION_MAP);
        return selectionBased.stream().map(this::addPrefix).collect(toSet());
    }

}

File: mns-core/src/main/java/cz/fb/manaus/core/category/categorizer/SelectionRegexpCategorizer.kt
package cz.fb.manaus.core.category.categorizer

import cz.fb.manaus.core.category.BetCoverage
import cz.fb.manaus.core.model.RealizedBet
import org.springframework.stereotype.Component

val SELECTION_MAP = mapOf(
        "draw" to compile("^The\\s+Draw$"),
        "yes" to compile("^Yes$"),
        "no" to compile("^No$"))

@Component
class SelectionRegexpCategorizer(
        private val regexpCategoryService: RegexpCategoryService
) : RealizedBetCategorizer {

    override fun getCategories(realizedBet: RealizedBet, coverage: BetCoverage): Set<String> {
        val selectionName = realizedBet.settledBet.selectionName
        return getCategories(selectionName)
    }

    internal fun getCategories(selectionName: String): Set<String> {
        val selectionBased = regexpCategoryService.getCategories(selectionName, SELECTION_MAP)
        return selectionBased.map { "selectionRegexp_$it" }.toSet()
    }
}
